appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/pool/datastructure/RWLockDataStructure.java

```diff
@@ -59,52 +62,52 @@
  */
 public class RWLockDataStructure implements DataStructure {
 
-    private ResourceHandler handler;
-    private ResourceSelectionStrategy strategy;
+    private final ResourceHandler handler;
     private int maxSize;
 
-    private final ArrayList<ResourceHandle> resources;
-    private ReentrantReadWriteLock reentrantLock = new ReentrantReadWriteLock();
-    private ReentrantReadWriteLock.ReadLock readLock = reentrantLock.readLock();
-    private ReentrantReadWriteLock.WriteLock writeLock = reentrantLock.writeLock();
+    private final List<ResourceHandle> resources;
+    private final ReentrantReadWriteLock reentrantLock = new ReentrantReadWriteLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = reentrantLock.readLock();
+    private final ReentrantReadWriteLock.WriteLock writeLock = reentrantLock.writeLock();
 
-    protected final static Logger _logger =
-            LogDomains.getLogger(RWLockDataStructure.class,LogDomains.RSR_LOGGER);
+    protected static final Logger _logger = LogDomains.getLogger(RWLockDataStructure.class,LogDomains.RSR_LOGGER);
 
-    public RWLockDataStructure(String parameters, int maxSize,
-                                              ResourceHandler handler, String strategyClass) {
-        resources = new ArrayList<ResourceHandle>((maxSize > 1000) ? 1000 : maxSize);
+    public RWLockDataStructure(int maxSize, ResourceHandler handler) {
+        resources = new ArrayList<>(Math.min(maxSize, 1000));
         this.maxSize = maxSize;
         this.handler = handler;
-        initializeStrategy(strategyClass);
         if(_logger.isLoggable(Level.FINEST)) {
             _logger.log(Level.FINEST, "pool.datastructure.rwlockds.init");
         }
     }
 
-    private void initializeStrategy(String strategyClass) {
-        //TODO
-    }
-
     /**
      * {@inheritDoc}
      */
     public int addResource(ResourceAllocator allocator, int count) throws PoolingException {
         int numResAdded = 0;
-        writeLock.lock();
-        //for now, coarser lock. finer lock needs "resources.size() < maxSize()" once more.
+        readLock.lock();
         try {
             for (int i = 0; i < count && resources.size() < maxSize; i++) {
```

Comment:
Nitpicking: Is following construction possible?

```java
            for (int i = 0; i < count && doLockSecured(this::isNotFull, readLock); i++) {
                writeLock.lock();
                try {
                    if (isNotFull()) {
                        ResourceHandle handle = handler.createResource(allocator);
                        resources.add(handle);
                        numResAdded++;
                    }                  
                } catch (Exception e) {
                    PoolingException pe = new PoolingException(e.getMessage());
                    pe.initCause(e);
                    throw pe;
                } finally {
                    writeLock.unlock();
                }
             }
```

At first glance it improves writelock downgrade, which is in this case possible unnecessary. Am I right?

Reply:
2nd version:

```java
 for (int i = 0; i < count && doLockSecured(this::isNotFull, readLock); i++) {
         numResourcesAdded += doLockSecured(this::addResourceInternal, writeLock);
 }
```

Reply:
In IMHO it would be best not to create the resource within the write lock. But rather create it before without lock then try to add it write locked and if this does not succeed destroy it again.
```
 for (int i = 0; i < count && doLockSecured(this::isNotFull, readLock); i++) {
         final  ResourceHandle handle = handler.createResource(allocator);
         boolean added = doLockSecured(() -> addResourceInternal(handle), writeLock);
         if (added) {
             numResourcesAdded++;
         } else {
             handler.removeResource(handle);
         }
 }
``` 
This way the DS is not locked during connection time resulting in higher throughput overall.

Reply:
I agree with @svendiedrichsen, as in this way, the resource's creation outside the write lock's scope will lead to better performance, especially when a batch of connections is being requested by the associated pool to each resource.